File: packages/core/src/bands.ts
import bandsData from "./bands.json";

/**
 * Band definitions are loaded from bands.json, which is the canonical table shared
 * with the Python inference service (services/inference/app/bands.py reads the same
 * file). Never redeclare these thresholds — change bands.json instead.
 *
 * Non-negotiable: never collapse this to a binary FAKE/REAL verdict anywhere
 * that consumes this table (web, extension, API responses).
 */
export interface BandDefinition {
  readonly id: string;
  readonly min: number;
  readonly max: number;
  readonly label: string;
  readonly copy: string;
}

export const BAND_DEFINITIONS: readonly BandDefinition[] = bandsData.bands;

export type BandId = "low" | "weak" | "mixed" | "strong" | "very_strong";

export const BAND_IDS = BAND_DEFINITIONS.map((b) => b.id) as [BandId, ...BandId[]];

/**
 * Maps a calibrated score in [0, 1] to its band definition.
 * Upper bound of the last band is inclusive; every other upper bound is exclusive,
 * so a score sitting exactly on a boundary (e.g. 0.45) falls into the higher band.
 */
export function scoreToBand(score: number): BandDefinition {
  if (score < 0 || score > 1 || Number.isNaN(score)) {
    throw new RangeError(`score must be a finite number in [0, 1], got ${score}`);
  }

  const band = BAND_DEFINITIONS.find((b, i) => {
    const isLast = i === BAND_DEFINITIONS.length - 1;
    return score >= b.min && (isLast ? score <= b.max : score < b.max);
  });

  if (!band) {
    throw new RangeError(`no band matched score ${score}`);
  }

  return band;
}

export const REPORT_FOOTER_DISCLAIMER: string = bandsData.report_footer_disclaimer;

File: services/inference/app/bands.py
"""Score -> band mapping.

The band table itself lives in ``packages/core/src/bands.json`` and is shared with
the TypeScript side. This module reads that file rather than restating the
thresholds, so there is exactly one place a threshold can be changed.

Non-negotiable: never collapse a band into a binary FAKE/REAL verdict.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass
from functools import lru_cache
from pathlib import Path


@dataclass(frozen=True)
class Band:
    id: str
    min: float
    max: float
    label: str
    copy: str


def _locate_bands_json() -> Path:
    """Find packages/core/src/bands.json.

    Explicit override wins, so the file can be copied next to the service in a
    container image where the monorepo layout isn't preserved.
    """
    override = os.environ.get("VERIFRAME_BANDS_JSON")
    if override:
        path = Path(override)
        if not path.is_file():
            raise FileNotFoundError(f"VERIFRAME_BANDS_JSON points at a missing file: {path}")
        return path

    here = Path(__file__).resolve()
    candidates = [
        # monorepo layout: services/inference/app/bands.py -> repo root
        here.parents[3] / "packages" / "core" / "src" / "bands.json",
        # container layout: bands.json copied alongside the app package
        here.parent / "bands.json",
    ]
    for candidate in candidates:
        if candidate.is_file():
            return candidate

    raise FileNotFoundError(
        "could not locate bands.json; set VERIFRAME_BANDS_JSON to its path. "
        f"Looked in: {', '.join(str(c) for c in candidates)}"
    )


@lru_cache(maxsize=1)
def _load() -> tuple[tuple[Band, ...], str]:
    data = json.loads(_locate_bands_json().read_text(encoding="utf-8"))
    bands = tuple(
        Band(id=b["id"], min=b["min"], max=b["max"], label=b["label"], copy=b["copy"])
        for b in data["bands"]
    )
    return bands, data["report_footer_disclaimer"]


def band_definitions() -> tuple[Band, ...]:
    return _load()[0]


def report_footer_disclaimer() -> str:
    return _load()[1]


def score_to_band(score: float) -> Band:
    """Map a calibrated score in [0, 1] to its band.

    Lower bounds are inclusive and upper bounds exclusive, except for the final
    band whose upper bound is inclusive. A score exactly on a boundary therefore
    lands in the higher band -- matching ``scoreToBand`` in packages/core.
    """
    if score != score or score < 0.0 or score > 1.0:  # NaN-safe range check
        raise ValueError(f"score must be a finite number in [0, 1], got {score!r}")

    bands = band_definitions()
    last_index = len(bands) - 1
    for i, band in enumerate(bands):
        upper_ok = score <= band.max if i == last_index else score < band.max
        if score >= band.min and upper_ok:
            return band

    raise ValueError(f"no band matched score {score!r}")
